drivers/ethernet/eth_w5500.c

```diff
@@ -275,39 +275,73 @@ static void w5500_rx(const struct device *dev)
 	w5500_command(dev, S0_CR_RECV);
 }
 
+static void w5500_update_link_status(const struct device *dev)
+{
+	uint8_t phycfgr;
+	struct w5500_runtime *ctx = dev->data;
+	int res = w5500_spi_read(dev, W5500_PHYCFGR, &phycfgr, 1);
+
+	if (res == 0) {
+		if (phycfgr & 0x01) {
+			if (ctx->link_up != true) {
+				LOG_INF("%s: Link up", dev->name);
+				ctx->link_up = true;
+				net_eth_carrier_on(ctx->iface);
+			}
+		} else {
+			if (ctx->link_up != false) {
+				LOG_INF("%s: Link down", dev->name);
+				ctx->link_up = false;
+				net_eth_carrier_off(ctx->iface);
+			}
+		}
+	}
+}
+
 static void w5500_thread(void *p1, void *p2, void *p3)
 {
 	ARG_UNUSED(p2);
 	ARG_UNUSED(p3);
 
 	const struct device *dev = p1;
 	uint8_t ir;
+	int res;
 	struct w5500_runtime *ctx = dev->data;
 	const struct w5500_config *config = dev->config;
 
 	while (true) {
-		k_sem_take(&ctx->int_sem, K_FOREVER);
+		res = k_sem_take(&ctx->int_sem, K_MSEC(CONFIG_ETH_W5500_LINK_STATUS_CHECK_MS));
```

Comment:
This will leave the thread running all the time even when there is no communication or even when the cable is not connected. I doubt PHY status checking with polling mode is really required here.

Reply:
I've tried to optimize it, with no luck so far. The idea was:
```c
if (ctx->link_up) {
	/* check link periodically */
	res = k_sem_take(&ctx->int_sem, K_MSEC(CONFIG_ETH_W5500_LINK_STATUS_CHECK_MS));
} else {
	/* link is down, wait for interrupt indefinitely */
	res = k_sem_take(&ctx->int_sem, K_FOREVER);
}
```

However, this does not detect reliably the link up event. Sometimes it just waits seemingly indefinitely to get an interrupt. I checked with the [W5500 datasheet](https://www.wiznet.hk/index.php?controller=attachment&id_attachment=5), pages 35 (IR Interrupt Register) and 48 (Sn_IR), looks like there are no flags related to the link status.

The approach used in this PR is identical to `stm32_eth_hal`, and it works reliably:
https://github.com/zephyrproject-rtos/zephyr/blob/7cfc8b3340b6cd0992034b1922bfa478bed7c730/drivers/ethernet/eth_stm32_hal.c#L890-L891

Reply:
@parthitce WDYT?

Reply:
> @parthitce WDYT?

Couldn't find time yet. Will be able to check the datasheet stuff tomorrow.

Reply:
The change works really well with my setup, I'd love to have it integrated.